lib/date_range.rb

```diff
@@ -0,0 +1,73 @@
+require "date"
+
+module HotelSystem
+  class DateRange
+    attr_reader :start_year, :start_month, :start_day, :num_nights
+
+    def initialize(start_year:, start_month:, start_day:, num_nights: nil)
+      self.class.valid_date_entry?(start_year, start_month, start_day)
+      @start_year = start_year
+      @start_month = start_month
+      @start_day = start_day
+      @num_nights = num_nights
+    end
+
+    def self.valid_date_entry?(year, month, day)
+      if year.digits.length != 4
+        raise ArgumentError, "Please enter 4 digits for the year."
+      elsif !Date.valid_date?(year, month, day)
+        raise ArgumentError, "Please enter a valid date with 4 digits for the year and 1 or 2 digits for both the month and day."
+      end
+    end
+
+    def date_list
+      dates_array = []
+      start_date = Date.new(start_year, start_month, start_day)
+      dates_array << start_date
+      if num_nights 
+        i = 1
+        num_nights.times do
+          date = start_date + i
+          dates_array << date
+          i += 1
```

Comment:
`.times` has an optional block parameter that contains the loop counter. In that case this could could be:

```ruby
num_nights.times do |i|
  dates_array << start_date + i
end
```